MoreLinq.Test/OrderedMergeTest.cs

```diff
@@ -0,0 +1,398 @@
+        }
+
+        [Test]
+        public void WhenThereIsACollisionShouldUseBothSelectorToChooseFromFirstOfSecondCollection()
+        {
+            var firstElement = new Version(3, 0);
+            var secondElement = new Version(3, 0);
+
+            var first = new[] { firstElement };
+            var second = new[] { secondElement };
+            Version First(Version f, Version _) => f;
+            Version Second(Version _, Version s) => s;
+
+            Version Id(Version id) => id;
+
+            var firstMerge = MoreEnumerable.OrderedMerge(
+                first: first,
+                second: second,
+                firstKeySelector: Id,
+                secondKeySelector: Id,
+                firstSelector: Id,
+                secondSelector: Id,
+                bothSelector: First,
+                comparer: null);
+            Assert.That(firstMerge.First(), Is.SameAs(firstElement), "Should have returned First");
+
+            var secondMerge = MoreEnumerable.OrderedMerge(
+                first: first,
+                second: second,
+                firstKeySelector: Id,
+                secondKeySelector: Id,
+                firstSelector: Id,
+                secondSelector: Id,
+                bothSelector: Second,
+                comparer: null);
+            Assert.That(secondMerge.First(), Is.SameAs(secondElement), "Should have returned Second");
+        }
+
+        [Test]
+        public void ShouldBeAbleToSelectKeyOfFirstAndSecondCollection()
+        {
+            var firstElement = new Version(2, 0);
+            var secondElement = new Version(3, 0);
+
+            var first = new[] { firstElement };
+            var second = new[] { secondElement };
+            int FirstKeySelector(Version version) => 2 * version.Major;
+            int SecondKeySelector(Version version) => version.Major;
+
+            Version Id(Version id) => id;
+            Version First(Version f, Version _) => f;
+
+            var merged = MoreEnumerable.OrderedMerge(
+                first: first,
+                second: second,
+                firstKeySelector: FirstKeySelector,
+                secondKeySelector: SecondKeySelector,
+                firstSelector: Id,
+                secondSelector: Id,
+                bothSelector: First,
+                comparer: null);
+
+            Assert.That(merged, Is.EqualTo(new[] { secondElement, firstElement }));
+        }
+
+        [Test]
+        public void FirstAndSecondCanBeOfDifferentInputTypesWithASharedOutputType()
+        {
+            int[] first = { 1, 4, 5 };
+            string[] second = { "2", "3", "6" };
+
+            int FirstKeySelector(int value) => value;
+            int SecondKeySelector(string value) => int.Parse(value);
+            int FirstSelector(int value) => value;
+            int SecondSelector(string value) => int.Parse(value);
+
+            int First(int f, string _) => f;
+
+            var merged = MoreEnumerable.OrderedMerge(
```

Comment:
Seems excessive to have named arguments and local functions, all reading the same. You have to move your eyes around and _mentally_ inline each definition. I would recommend in-lining all local functions such that it reads simple:

```c#
            var merged =
                first.OrderedMerge(second,
                    firstKeySelector : value => value,
                    secondKeySelector: int.Parse,
                    firstSelector    : value => value,
                    secondSelector   : int.Parse,
                    bothSelector     : (f, _) => f,
                    comparer         : null);
```

It's fine to name the argument, but I'd skip it for `second`. I would review this for all similar patterns.